subsys/logging/log_backend_adsp_hda.c

```diff
@@ -245,17 +333,81 @@ static int format_set(const struct log_backend *const backend, uint32_t log_type
 	return 0;
 }
 
-static volatile uint32_t counter;
+
+/* Create a little output section for locally formatting data */
+#define LOCALS 4
+static atomic_t local_used = ATOMIC_INIT(0);
+uint8_t local_buf[LOCALS][128];
+struct log_output_control_block ctrl_block;
+struct log_output outputs[LOCALS] = {
+	{
+		.func = hda_log_out,
+		.control_block = &ctrl_block,
+		.buf  = &local_buf[0][0],
+		.size = sizeof(local_buf),
+	},
+	{
+		.func = hda_log_out,
+		.control_block = &ctrl_block,
+		.buf  = &local_buf[1][0],
+		.size = sizeof(local_buf),
+	},
+	{
+		.func = hda_log_out,
+		.control_block = &ctrl_block,
+		.buf  = &local_buf[2][0],
+		.size = sizeof(local_buf),
+	},
+	{
+		.func = hda_log_out,
+		.control_block = &ctrl_block,
+		.buf  = &local_buf[3][0],
+		.size = sizeof(local_buf),
+	}
+};
 
 static void process(const struct log_backend *const backend,
 		union log_msg_generic *msg)
 {
 	ARG_UNUSED(backend);
+
 	uint32_t flags = log_backend_std_get_flags();
 
 	log_format_func_t log_output_func = log_format_func_t_get(log_format_current);
 
-	log_output_func(&log_output_adsp_hda, &msg->log, flags);
+
+	/* Find a free output to write formatting into
+	 *
+	 * Spin tightly here on a few outputs that enable
+	 * concurent formatting to occur. If done in a thread
+	 * this loop is safely preemptible.
+	 *
+	 * For ISRs the normal "panic" output path is provided
+	 * with a spin lock.
+	 */
+	if (k_is_in_isr()) {
+		k_spinlock_key_t key = k_spin_lock(&hda_log_fmt_lock);
+
+		log_output_func(&log_output_adsp_hda, &msg->log, flags);
+
+		k_spin_unlock(&hda_log_fmt_lock, key);
+
+		return;
+	}
+
+	int32_t local = -1;
+
+	do {
+		for (int i = 0; i < LOCALS; i++) {
+			if (!atomic_test_and_set_bit(&local_used, i)) {
+				local = i;
+			}
+		}
+	} while (local == -1);
+
+	log_output_func(&outputs[local], &msg->log, flags);
```

Comment:
Ok this is quite clever and probably something we need to copy to other backends as well. I wonder if we could share this code somehow so it is easier to reuse. It seems this is needed by any backend that runs on multicore systems to avoid blocking interrupts for long periods of time.

Reply:
It did require a bit more finesse than this it turns out, but it works really really well with LOG_MODE_IMMEDIATE. In effect every core independently deals with formatting now and the throughput of the logger went from something like 8-20k/s to ~500k/s with 8 threads on cavs25.